pkg/search/postgres/store_cache.go

```diff
@@ -0,0 +1,394 @@
+	}
+	dbErr := c.underlyingStore.DeleteMany(ctx, filteredIDs)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.RemoveMany)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	for _, id := range filteredIDs {
+		delete(c.cache, id)
+	}
+	return nil
+}
+
+// Exists tells whether the ID exists in the store.
+func (c *cachedStore[T, PT]) Exists(ctx context.Context, id string) (bool, error) {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Exists)
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	obj, found := c.cache[id]
+	if !found {
+		return false, nil
+	}
+	return c.isReadAllowed(ctx, obj), nil
+}
+
+// Count returns the number of objects in the store.
+func (c *cachedStore[T, PT]) Count(ctx context.Context) (int, error) {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Count)
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	count := 0
+	walkErr := c.walkCacheNoLock(ctx, func(obj PT) error {
+		count++
+		return nil
+	})
+	if walkErr != nil {
+		return 0, walkErr
+	}
+	return count, nil
+}
+
+// Get returns the object, if it exists from the store.
+func (c *cachedStore[T, PT]) Get(ctx context.Context, id string) (PT, bool, error) {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Get)
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	obj, found := c.cache[id]
+	if !found {
+		return nil, false, nil
+	}
+	if !c.isReadAllowed(ctx, obj) {
+		return nil, false, nil
+	}
+	return obj, true, nil
+}
+
+// GetMany returns the objects specified by the IDs from the store as well as the index in the missing indices slice.
+func (c *cachedStore[T, PT]) GetMany(ctx context.Context, identifiers []string) ([]PT, []int, error) {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.GetMany)
+	if len(identifiers) == 0 {
+		return nil, nil, nil
+	}
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	results := make([]PT, 0, len(identifiers))
+	misses := make([]int, 0, len(identifiers))
+	for ix, id := range identifiers {
+		obj, found := c.cache[id]
+		if !found {
+			misses = append(misses, ix)
+			continue
+		}
+		if !c.isReadAllowed(ctx, obj) {
+			misses = append(misses, ix)
+			continue
+		}
+		results = append(results, obj)
+	}
+	return results, misses, nil
+}
+
+// Walk iterates over all the objects in the store and applies the closure.
+func (c *cachedStore[T, PT]) Walk(ctx context.Context, fn func(obj PT) error) error {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	return c.walkCacheNoLock(ctx, fn)
+}
+
+// GetByQuery returns the objects from the store matching the query.
+func (c *cachedStore[T, PT]) GetByQuery(ctx context.Context, query *v1.Query) ([]*T, error) {
+	return c.underlyingStore.GetByQuery(ctx, query)
+}
+
+// DeleteByQuery removes the objects from the store based on the passed query.
+func (c *cachedStore[T, PT]) DeleteByQuery(ctx context.Context, query *v1.Query) error {
+	objs, dbFetchErr := c.underlyingStore.GetByQuery(ctx, query)
```

Comment:
Hmm, this seems pretty expensive. How hard would it be to run the query and then delete. I guess another option would be to modify the store to return the ids that were deleted by the query which seems like more work 🤔

Reply:
The other option would be to potentially force the application code to do this, so we'd effectively not support delete by query on a cached store?

Reply:
The `DeleteByQuery` case was heavily reworked in #8975